WebScraping/tests/unit/test_parser.py

```diff
@@ -1,4 +1,226 @@
 from WebScraping.functions.parser.lambda_function import *
+from unittest.mock import MagicMock, patch
+from bs4 import BeautifulSoup
+from botocore.stub import Stubber
+import os
+import pytest
 
-def test():
-    assert True == True
\ No newline at end of file
+FILE_PATH = "WebScraping.functions.parser.lambda_function"
+
+def test_lambda_handler():
+    s3_href_object = MagicMock()
+    s3_href_object.bucket_name = BUCKET_NAME
+    s3_href_object.key = KEY_PATH + "href_list.json"
+    s3_href_object.get.return_value = {"Body": MagicMock(read=lambda: "[]")}
+
+    s3_classes_object = MagicMock()
+    s3_classes_object.bucket_name = BUCKET_NAME
+    s3_classes_object.key = KEY_PATH + "classes.json"
+    s3_classes_object.get.return_value = {"Body": MagicMock(read=lambda: "")}
+
+    with patch(f"{FILE_PATH}.get_s3_object", side_effect=[s3_href_object, s3_classes_object]):
+        result = lambda_handler(event={}, context={})
+        s3_href_object.get.assert_called_once_with()
+        s3_classes_object.put.assert_called_once_with(Body=b"{}")
+        assert result == "classes JSON written to s3"
+
+
+def test_get_data():
+    href_dict = {"href": "test-course-href", "also_register": "test-also-register-str"}
+    test_html = "<html><body>Mocked HTML</body></html>"
+    expected_url = BASE_URL + href_dict["href"]
+    expected_bs4 = BeautifulSoup(test_html, "html.parser")
+
+    with patch("urllib3.PoolManager.request") as mock_pool_manager:
+        mock_response = MagicMock()
+        mock_response.data = test_html.encode()
+        mock_pool_manager.return_value = mock_response
+
+        with patch(f"{FILE_PATH}.parse_data") as mock_parser:
+            get_data(href_dict)
+            mock_pool_manager.assert_called_once()
+            mock_pool_manager.assert_called_once_with("GET", expected_url)
+            mock_parser.assert_called_once_with(expected_bs4, href_dict["also_register"])
+
+
+def test_get_s3_object():
+    s3_client = boto3.client("s3")
```

Comment:
Great job on the tests, just a quick question: when running this line locally (ie not in AWS) does it not require you to provide the region? When I had boto3.client("dynamodb") in my test it caused an error since no region was set.

Reply:
When i was running it locally it did not require me to set the region, i can set the region in case so it does not fail when the pipeline runs